### PR TITLE
add an endpoint to recieve a query string of programme membership ids…

### PR DESCRIPTION
**Pull request by EdwardBarclay**
… and return an array of programmeMembershipSummaryDto for each id

PR https://github.com/Health-Education-England/TIS-ADMINS-UI/pull/1414 is dependent on this PR being in place before merging.

TIS21-6276

### CODE REVIEW

**Review by JosephKelly — DISMISSED** [review]
This absolutely needs some tests, TDD might help you work through it.  An integration test should give you what you need without the need for a unit test as well.

File: tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/ProgrammeMembershipResource.java
Comment (JosephKelly) [inline tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/ProgrammeMembershipResource.java:182]: I forgot to say, I think there's probably a more appropriate method on the service to use for this search rather than looping over each id separately to do a `findOne`.

File: tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/ProgrammeMembershipResource.java
Comment (EdwardBarclay) [reply tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/ProgrammeMembershipResource.java:175]: I am updating it to be more inline for REST principles using query parameters which are more appropriate for multiple paramaters

**Comment by EdwardBarclay** [comment]
> This absolutely needs some tests, TDD might help you work through it. An integration test should give you what you need without the need for a unit test as well.

Test was delayed because the project is not allowing me to test claiming missing queues and flayway files, definitely adding a test TDD would have been preferable

File: tcs-api/src/main/java/com/transformuk/hee/tis/tcs/api/dto/ProgrammeMembershipSummaryDTO.java
Comment (ReubenRobertsHEE) [inline tcs-api/src/main/java/com/transformuk/hee/tis/tcs/api/dto/ProgrammeMembershipSummaryDTO.java:13]: Do we mean 'programmeName' here? Otherwise I'm not sure what this field would contain, so it should have a comment if its not a std. ProgrammeMembership field I think.

File: tcs-api/src/main/java/com/transformuk/hee/tis/tcs/api/dto/ProgrammeMembershipSummaryDTO.java
Comment (ReubenRobertsHEE) [inline tcs-api/src/main/java/com/transformuk/hee/tis/tcs/api/dto/ProgrammeMembershipSummaryDTO.java:12]: Not sure what field is meant here: is it the uuid? If so, it should probably be called 'uuid' (the 'id' field in the ProgrammeMembership entity is deprecated).

File: tcs-persistence/src/main/java/com/transformuk/hee/tis/tcs/service/repository/ProgrammeMembershipRepository.java
Comment (ReubenRobertsHEE) [inline tcs-persistence/src/main/java/com/transformuk/hee/tis/tcs/service/repository/ProgrammeMembershipRepository.java:53]: Given there is already something which takes a `Set<UUID>` and returns the same content, is it not possible to reuse that? It feels a little awkward to duplicate the same functionality here.

File: tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/ProgrammeMembershipResource.java
Comment (ReubenRobertsHEE) [inline tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/ProgrammeMembershipResource.java:172]: Unless there is a good reason for the change, probably best to leave the debug logging 😁

File: tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/ProgrammeMembershipResource.java
Comment (ReubenRobertsHEE) [inline tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/ProgrammeMembershipResource.java:74]: Is there a reason why this isn't put into the service.mappers package? It feels a bit odd to have it included here.

File: tcs-persistence/src/main/java/com/transformuk/hee/tis/tcs/service/repository/ProgrammeMembershipRepository.java
Comment (EdwardBarclay) [reply tcs-persistence/src/main/java/com/transformuk/hee/tis/tcs/service/repository/ProgrammeMembershipRepository.java:53]: the reason I have it as a List alternative is because we can receive duplicate ids such as if a form A and B are for the same programme, a list allows duplicates where as a Set is unique only

Comment (ReubenRobertsHEE) [reply tcs-persistence/src/main/java/com/transformuk/hee/tis/tcs/service/repository/ProgrammeMembershipRepository.java:53]: Ah, ok. Do you have another PR that I can use to see the whole journey? It still feels odd to me to need to include duplicates, since the ProgrammeMembershipSummaryDTO content will be identical for these, so I would have thought that a single entry could be 'duplicated' as necessary when producing the forms listing, but maybe I'm misunderstanding the process?

File: tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/ProgrammeMembershipResource.java
Comment (ReubenRobertsHEE) [reply tcs-service/src/main/java/com/transformuk/hee/tis/tcs/service/api/ProgrammeMembershipResource.java]: Hi Ed, this looks good! Do you think you could implement a test or two to demonstrate what happens when (some of) the UUIDs are malformed, or don't match to anything in the DB? Do we get nothing, or a partial listing excluding those bad parameters?

**Review by ReubenRobertsHEE — APPROVED** [review]

**Review by ReubenRobertsHEE — DISMISSED** [review]
Do get approval from the admin team before merging please :)

**Review by ReubenRobertsHEE — APPROVED** [review]

**Pull request merged**
